base/checked.jl

```diff
@@ -0,0 +1,333 @@
+
+function checked_abs(x::SignedInt)
+    r = ifelse(x<0, -x, x)
+    r<0 && throw(OverflowError())
+    r
+ end
+checked_abs(x::UnsignedInt) = x
+
+"""
+    Base.checked_add(x, y)
+
+Calculates `x+y`, checking for overflow errors where applicable.
+
+The overflow protection may impose a perceptible performance penalty.
+"""
+function checked_add end
+
+function checked_add{T<:SignedInt}(x::T, y::T)
+    box(T, checked_sadd_int(unbox(T,x), unbox(T,y)))
+end
+function checked_add{T<:UnsignedInt}(x::T, y::T)
+    box(T, checked_uadd_int(unbox(T,x), unbox(T,y)))
+end
+function checked_add{T<:BrokenSignedInt}(x::T, y::T)
+    r = x + y
+    # x and y have the same sign, and the result has a different sign
+    (x<0) == (y<0) != (r<0) && throw(OverflowError())
+    r
+end
+function checked_add{T<:BrokenUnsignedInt}(x::T, y::T)
+    # x + y > typemax(T)
+    # Note: ~y == -y-1
+    x > ~y && throw(OverflowError())
+    x + y
+end
+
+# Handle multiple arguments
+checked_add(x) = x
+checked_add{T}(x1::T, x2::T, x3::T) =
+    checked_add(checked_add(x1, x2), x3)
+checked_add{T}(x1::T, x2::T, x3::T, x4::T) =
+    checked_add(checked_add(x1, x2), x3, x4)
+checked_add{T}(x1::T, x2::T, x3::T, x4::T, x5::T) =
+    checked_add(checked_add(x1, x2), x3, x4, x5)
+checked_add{T}(x1::T, x2::T, x3::T, x4::T, x5::T, x6::T) =
+    checked_add(checked_add(x1, x2), x3, x4, x5, x6)
+checked_add{T}(x1::T, x2::T, x3::T, x4::T, x5::T, x6::T, x7::T) =
+    checked_add(checked_add(x1, x2), x3, x4, x5, x6, x7)
+checked_add{T}(x1::T, x2::T, x3::T, x4::T, x5::T, x6::T, x7::T, x8::T) =
+    checked_add(checked_add(x1, x2), x3, x4, x5, x6, x7, x8)
+
+"""
+    Base.checked_sub(x, y)
+
+Calculates `x-y`, checking for overflow errors where applicable.
+
+The overflow protection may impose a perceptible performance penalty.
+"""
+function checked_sub end
+
+function checked_sub{T<:SignedInt}(x::T, y::T)
+    box(T, checked_ssub_int(unbox(T,x), unbox(T,y)))
+end
+function checked_sub{T<:UnsignedInt}(x::T, y::T)
+    box(T, checked_usub_int(unbox(T,x), unbox(T,y)))
+end
+function checked_sub{T<:BrokenSignedInt}(x::T, y::T)
+    r = x - y
+    # x and y have different signs, and the result has a different sign than x
+    (x<0) != (y<0) == (r<0) && throw(OverflowError())
+    r
+end
+function checked_sub{T<:BrokenUnsignedInt}(x::T, y::T)
+    # x - y < 0
+    x < y && throw(OverflowError())
+    x - y
+end
+
+"""
+    Base.checked_mul(x, y)
+
+Calculates `x*y`, checking for overflow errors where applicable.
+
+The overflow protection may impose a perceptible performance penalty.
+"""
+function checked_mul end
+
+function checked_mul{T<:SignedInt}(x::T, y::T)
+    box(T, checked_smul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:UnsignedInt}(x::T, y::T)
+    box(T, checked_umul_int(unbox(T,x), unbox(T,y)))
+end
+function checked_mul{T<:BrokenSignedIntMul}(x::T, y::T)
+    r = nextwide(x) * nextwide(y)
+    !(typemin(T) <= r <= typemax(T)) && throw(OverflowError())
+    T(r)
```

Comment:
It might be faster here to use `r % T`, as that doesn't do checking (it's also possible that LLVM is able to optimise the branch out).